src/WebApplication.php

```diff
@@ -97,115 +111,122 @@ public function doExecute()
+			'accessToken'       => $input->getString($this->get('webservices.oauth2_token_param_name', 'access_token')),
+			'id'                => $input->getString('id'),
+			'absoluteHrefs'     => $input->getBool('absoluteHrefs', true),
+		);
+
+		$rendererOptions = [
+			'charset'		=> 'utf-8',
+			'language'		=> 'en-GB',
+			'direction'		=> 'ltr',
+			'link'			=> '',
+			'base'			=> '',
+			'absoluteHrefs'	=> $input->getBool('absoluteHrefs', true),
+			'uriParams'		=> [],
+			'resourceName'	=> $resourceName,
+		];
 
 		try
 		{
-			$this->clearHeaders();
-			$webserviceClient = $input->getString('webserviceClient', 'administrator');
-			$optionName       = $input->getString('option');
-			$optionName       = strpos($optionName, 'com_') === 0 ? substr($optionName, 4) : $optionName;
-			$viewName         = $input->getString('view');
-			$version          = $input->getString('webserviceVersion');
-
-			$token = $input->getString($this->get('webservices.oauth2_token_param_name', 'access_token'));
-			$apiName = ucfirst($apiName);
-			$method  = strtoupper($input->getMethod());
-			$task    = $this->getTask();
-			$data    = $this->getPostedData();
-			$dataGet = $input->getArray();
-
-			$options = array(
-				'api'               => $apiName,
-				'optionName'        => $optionName,
-				'viewName'          => $viewName,
-				'webserviceVersion' => $version,
-				'webserviceClient'  => $webserviceClient,
-				'method'            => $method,
-				'task'              => $task,
-				'data'              => $data,
-				'dataGet'           => $dataGet,
-				'accessToken'       => $token,
-				'format'            => $input->getString('format', $this->get('webservices.webservices_default_format', 'hal')),
-				'id'                => $input->getString('id'),
-				'absoluteHrefs'     => $input->getBool('absoluteHrefs', true),
-			);
-
-			$apiClass = 'Joomla\\Webservices\\Api\\' . $apiName . '\\' . $apiName;
-
-			if (!class_exists($apiClass))
-			{
-				throw new \RuntimeException($text->sprintf('LIB_WEBSERVICES_API_UNABLE_TO_LOAD_API', $options['api']));
-			}
-
-			try
-			{
-				/** @var \Joomla\Webservices\Api\ApiBase $api */
-				$api = new $apiClass($this->getContainer(), new Registry($options));
-			}
-			catch (\RuntimeException $e)
-			{
-				throw new \RuntimeException($text->sprintf('LIB_WEBSERVICES_API_UNABLE_TO_CONNECT_TO_API', $e->getMessage()));
-			}
-
-			// Run the api task
-			$api->execute();
-
-			// Display output
-			$api->render();
+			// Instantiate a renderer, based on content negotiation, then defer to the API class.
+			$this->container
+				->registerServiceProvider(new RendererProvider($contentType, new Registry($rendererOptions)))
+				->registerServiceProvider(new ApiProvider(new Registry($options)))
+				->get('api')->execute()->render()
+				;
 		}
 		catch (\Exception $e)
 		{
+			// @TODO Generally, application errors should be handled by the API object,
+			// so the following code should probably be moved into the API class.
```

Comment:
Agreed.  The comment was really referring to the need to most handle errors in the API object so, for example, REST errors get returned in something like this: https://tools.ietf.org/html/draft-ietf-appsawg-http-problem-02  I think SOAP has its own error format too.  The code here should only be an absolute last resort bail out.